sw/device/lib/dif/dif_keymgr.c

```diff
@@ -5,3 +5,723 @@
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_INIT:
+      *state = kDifKeymgrStateInitialized;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_CREATOR_ROOT_KEY:
+      *state = kDifKeymgrStateCreatorRootKey;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_OWNER_INTERMEDIATE_KEY:
+      *state = kDifKeymgrStateOwnerIntermediateKey;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_OWNER_KEY:
+      *state = kDifKeymgrStateOwnerRootKey;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_DISABLED:
+      *state = kDifKeymgrStateDisabled;
+      break;
+    case KEYMGR_WORKING_STATE_STATE_VALUE_INVALID:
+      *state = kDifKeymgrStateInvalid;
+      break;
+    default:
+      return kDifKeymgrError;
+  }
+
+  return kDifKeymgrOk;
+}
+
+dif_keymgr_lockable_result_t dif_keymgr_generate_identity_seed(
+    const dif_keymgr_t *keymgr) {
+  if (keymgr == NULL) {
+    return kDifKeymgrLockableBadArg;
+  }
+
+  if (!is_idle(keymgr)) {
+    return kDifKeymgrLockableLocked;
+  }
+
+  start_operation(keymgr, (start_operation_params_t){
+                              .dest = KEYMGR_CONTROL_DEST_SEL_VALUE_NONE,
+                              .op = KEYMGR_CONTROL_OPERATION_VALUE_GENERATE_ID,
+                          });
+
+  return kDifKeymgrLockableOk;
+}
+
+dif_keymgr_lockable_result_t dif_keymgr_generate_versioned_key(
+    const dif_keymgr_t *keymgr, dif_keymgr_versioned_key_params_t params) {
+  if (keymgr == NULL) {
+    return kDifKeymgrLockableBadArg;
+  }
+
+  start_operation_params_t hw_op_params;
+  switch (params.dest) {
+    case kDifKeymgrVersionedKeyDestSw:
+      hw_op_params.dest = KEYMGR_CONTROL_DEST_SEL_VALUE_NONE;
+      hw_op_params.op = KEYMGR_CONTROL_OPERATION_VALUE_GENERATE_SW_OUTPUT;
+      break;
+    case kDifKeymgrVersionedKeyDestAes:
+      hw_op_params.dest = KEYMGR_CONTROL_DEST_SEL_VALUE_AES;
+      hw_op_params.op = KEYMGR_CONTROL_OPERATION_VALUE_GENERATE_HW_OUTPUT;
+      break;
+    case kDifKeymgrVersionedKeyDestHmac:
+      hw_op_params.dest = KEYMGR_CONTROL_DEST_SEL_VALUE_HMAC;
+      hw_op_params.op = KEYMGR_CONTROL_OPERATION_VALUE_GENERATE_HW_OUTPUT;
+      break;
+    case kDifKeymgrVersionedKeyDestKmac:
+      hw_op_params.dest = KEYMGR_CONTROL_DEST_SEL_VALUE_KMAC;
+      hw_op_params.op = KEYMGR_CONTROL_OPERATION_VALUE_GENERATE_HW_OUTPUT;
+      break;
+    default:
+      return kDifKeymgrLockableBadArg;
+  }
+
+  if (!is_idle(keymgr)) {
+    return kDifKeymgrLockableLocked;
+  }
+
+  // Set salt and version
+  mmio_region_memcpy_to_mmio32(keymgr->params.base_addr,
+                               KEYMGR_SALT_0_REG_OFFSET, params.salt,
+                               sizeof(params.salt));
+  mmio_region_write32(keymgr->params.base_addr, KEYMGR_KEY_VERSION_REG_OFFSET,
+                      params.version);
+
+  start_operation(keymgr, hw_op_params);
+
+  return kDifKeymgrLockableOk;
+}
+
+dif_keymgr_result_t dif_keymgr_sideload_clear_set_enabled(
```

Comment:
@tjaychen's earlier [comment](https://github.com/lowRISC/opentitan/pull/5463#issuecomment-796177458):
> if you keep it as 1, it will basically continually clear. So i guess it depends on whether we want to end up somewhere "random"-ish.

I can rename it to clear_start/clear_end in a follow-up PR if you want. I tried to leave this flexible using the names we already use.

Reply:
ack, start/end nomenclature sgtm. I guess from a hardware/security perspective it is nice to leave the interface as-is.